src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -135,4 +153,156 @@
+                } else {
+                    syncOptions.applyWarningCallback(format("%s with id: '%s' has no key set. Keys are"
+                        + " required for resource matching.", typeName, id));
+                }
+            });
+
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), query.get(), resourcePageConsumer)
+            .thenApply(result -> Optional.ofNullable(keyToIdCache.get(key)));
+    }
+
+    /**
+     * Given a {@link Set} of resource keys, this method fetches a set of all the resources, matching this given set of
+     * keys in the CTP project, defined in an injected {@link SphereClient}. A mapping of the key to the id
+     * of the fetched resources is persisted in an in-memory map.
+     *
+     * @param resourceKeys set of state keys to fetch matching states by
+     * @param query        a function to get query
+     * @param keyMapper    a function to get the key from the returned resource
+     * @return {@link CompletionStage}&lt;{@link Set}&lt;{@code U}&gt;&gt; in which the result of it's completion
+     *         contains a {@link Set} of all matching resources.
+     */
+    @Nonnull
+    CompletionStage<Set<U>> fetchMatchingResources(@Nonnull final Set<String> resourceKeys,
+                                                   @Nonnull final Supplier<Q> query,
+                                                   @Nonnull final Function<U, String> keyMapper) {
+        if (resourceKeys.isEmpty()) {
+            return CompletableFuture.completedFuture(Collections.emptySet());
+        }
+
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), query.get(), Function.identity())
+            .thenApply(fetchedResources -> fetchedResources
+                .stream()
+                .flatMap(List::stream)
+                .peek(resource -> keyToIdCache.put(keyMapper.apply(resource), resource.getId()))
+                .collect(Collectors.toSet()));
+    }
+
+    /**
+     * Given a resource key, this method fetches a resource that matches this given key in the CTP project defined in a
+     * potentially injected {@link SphereClient}. If there is no matching resource an empty {@link Optional} will be
+     * returned in the returned future. A mapping of the key to the id of the fetched category is persisted in an in
+     * -memory map.
+     *
+     * @param key       the key of the resource to fetch
+     * @param query     a function to get query
+     * @param keyMapper a function to get the key from the returned resource
+     * @return {@link CompletionStage}&lt;{@link Optional}&gt; in which the result of it's completion contains an
+     *         {@link Optional} that contains the matching {@code T} if exists, otherwise empty.
+     */
+    @Nonnull
+    CompletionStage<Optional<U>> fetchResource(@Nullable final String key,
+                                               @Nonnull final Supplier<Q> query,
```

Comment:
I don't understand why you need the query here as a parameter?

Doesn't this service already know that it should fetch the resource by key?

Reply:
Also, is there a reason why it's not used by the subclasses?

Reply:
I understood at first we are enhancing only BaseService class and to avoid to much changes at once we are leaving subclasses implementation as they are.
New issue will be created later to adapt to changes. PRs for [#409](https://github.com/commercetools/commercetools-sync-java/pull/420) and [#417](https://github.com/commercetools/commercetools-sync-java/pull/421) already are using this version of BaseService

Reply:
Alright, sounds good. I just wasn't 100% sure if it was by intension, so thought I'd ask :)

Reply:
Fixed: [e4c891d](https://github.com/commercetools/commercetools-sync-java/pull/419/commits/e4c891d1d4a14e94f7bf9deb70b3ad64e2da639f)